src/lib/array.jl

```diff
@@ -179,8 +188,9 @@ for (mapfunc,∇mapfunc) in [(:map,:∇map),(:pmap,:∇pmap),(:vmap,:∇vmap)]
     else
       ys, backs = unzip(ys_and_backs)
       ys, function (Δ)
-        Δf_and_args_zipped = $mapfunc((f, δ) -> f(δ), backs, Δ)
-        Δf_and_args = unzip(Δf_and_args_zipped)
+        # Apply pullbacks in reverse order. Needed for correctness if `f` is stateful.
+        Δf_and_args_zipped = $mapfunc((f, δ) -> f(δ), _tryreverse($mapfunc, backs, Δ)...) 
+        Δf_and_args = unzip(_tryreverse($mapfunc, Δf_and_args_zipped))
```

Comment:
Any particular reason for the refactoring? Might be better to split out the adjoint for pmap etc to just do the sync and pmap in the pullback.